cassandra3/src/main/java/zipkin/dependencies/cassandra3/CassandraDependenciesJob.java

```diff
@@ -0,0 +1,263 @@
+      sparkProperties.put("spark.cassandra.auth.password", getEnv("CASSANDRA_PASSWORD", ""));
+    }
+
+    /** When set, this indicates which jars to distribute to the cluster. */
+    public Builder jars(String... jars) {
+      this.jars = jars;
+      return this;
+    }
+
+    /** Keyspace to store dependency rowsToLinks. Defaults to "zipkin2" */
+    public Builder keyspace(String keyspace) {
+      this.keyspace = checkNotNull(keyspace, "keyspace");
+      return this;
+    }
+
+    /** Day (in epoch milliseconds) to process dependencies for. Defaults to today. */
+    public Builder day(long day) {
+      this.day = midnightUTC(day);
+      return this;
+    }
+
+    /** Ensures that logging is setup. Particularly important when in cluster mode. */
+    public Builder logInitializer(Runnable logInitializer) {
+      this.logInitializer = checkNotNull(logInitializer, "logInitializer");
+      return this;
+    }
+
+    /** Comma separated list of hosts / IPs part of Cassandra cluster. Defaults to localhost */
+    public Builder contactPoints(String contactPoints) {
+      this.contactPoints = contactPoints;
+      return this;
+    }
+
+    /** The local DC to connect to (other nodes will be ignored) */
+    public Builder localDc(@Nullable String localDc) {
+      this.localDc = localDc;
+      return this;
+    }
+
+    public CassandraDependenciesJob build() {
+      return new CassandraDependenciesJob(this);
+    }
+  }
+
+  final String keyspace;
+  final long day;
+  final String dateStamp;
+  final SparkConf conf;
+  @Nullable final Runnable logInitializer;
+
+  CassandraDependenciesJob(Builder builder) {
+    this.keyspace = builder.keyspace;
+    this.day = builder.day;
+    SimpleDateFormat df = new SimpleDateFormat("yyyy-MM-dd");
+    df.setTimeZone(TimeZone.getTimeZone("UTC"));
+    this.dateStamp = df.format(new Date(builder.day));
+    this.conf = new SparkConf(true).setMaster(builder.sparkMaster).setAppName(getClass().getName());
+    conf.set("spark.cassandra.connection.host", parseHosts(builder.contactPoints));
+    conf.set("spark.cassandra.connection.port", parsePort(builder.contactPoints));
+    if (builder.localDc != null) conf.set("connection.local_dc", builder.localDc);
+    if (builder.jars != null) conf.setJars(builder.jars);
+    for (Map.Entry<String, String> entry : builder.sparkProperties.entrySet()) {
+      conf.set(entry.getKey(), entry.getValue());
+    }
+    this.logInitializer = builder.logInitializer;
+  }
+
+  public void run() {
+    long microsLower = day * 1000;
+    long microsUpper = (day * 1000) + TimeUnit.DAYS.toMicros(1) - 1;
+
+    log.info(
+        "Running Dependencies job for {}: {} ≤ Span.timestamp {}",
+        dateStamp,
+        microsLower,
+        microsUpper);
+
+    SparkContext sc = new SparkContext(conf);
+    try {
+      JavaRDD<DependencyLink> links = javaFunctions(sc)
+          .cassandraTable(keyspace, "span")
+          .map(CassandraRowToSpan.INSTANCE)
```

Comment:
@llinder @michaelsembwever 

TL;DR; it would be more convenient if span.trace_id was always the lower 64 bits (right-most 16 chars) instead of 16 or 32 depending on length.

In our new schema, we store the 32chars as trace_id if strictTraceId=true. This creates a bit of a problem in spark as data with different trace IDs aren't "near eachother", so when scanning in mixed mode (default for spark), so we have decode and distribute more data.

For example, to use groupBy, we have to map every span into an object like you see above. I'm thinking it might be more practical to always have span.trace_id as the same as it was before traceIdLow, then partitions will work naturally and especially larger sites can process faster.. Reason is that for a long time there will be mixed bit length and unless we accommodate this one way or another, the dependency tree will be messed up.

wdyt?

https://github.com/datastax/spark-cassandra-connector/blob/master/doc/3_selection.md#grouping-rows-by-partition-key

Reply:
another way is we write the job in 2 different ways, then look at the `STRICT_TRACE_ID` variable. if `STRICT_TRACE_ID=false`, then we do the more expensive groupBy. I think I'll go that way honestly as people can always re-process a day's data if they found themselves in this situation

Reply:
Not sure I full understand the tradeoffs with only storing the lower 64 bits in `span.trace_id` but it sounds reasonable.   The second option seems like it would be acceptable as well but I'm probably biased on that decision because I'm using strictTraceId=true.